akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -693,3 +699,423 @@ private[akka] class BroadcastHub[T](bufferSize: Int) extends GraphStageWithMater
+ * of times, where each of the new materializations will receive their elements from the original [[Sink]].
+ */
+object PartitionHub {
+
+  /**
+   * FIXME write docs
+   *
+   * Creates a [[Sink]] that receives elements from its upstream producer and broadcasts them to a dynamic set
+   * of consumers. After the [[Sink]] returned by this method is materialized, it returns a [[Source]] as materialized
+   * value. This [[Source]] can be materialized an arbitrary number of times and each materialization will receive the
+   * broadcast elements form the ofiginal [[Sink]].
+   *
+   * Every new materialization of the [[Sink]] results in a new, independent hub, which materializes to its own
+   * [[Source]] for consuming the [[Sink]] of that materialization.
+   *
+   * If the original [[Sink]] is failed, then the failure is immediately propagated to all of its materialized
+   * [[Source]]s (possibly jumping over already buffered elements). If the original [[Sink]] is completed, then
+   * all corresponding [[Source]]s are completed. Both failure and normal completion is "remembered" and later
+   * materializations of the [[Source]] will see the same (failure or completion) state. [[Source]]s that are
+   * cancelled are simply removed from the dynamic set of consumers.
+   *
+   * @param bufferSize Buffer size used by the producer. Gives an upper bound on how "far" from each other two
+   *                   concurrent consumers can be in terms of element. If this buffer is full, the producer
+   *                   is backpressured. Must be a power of two and less than 4096.
+   */
+  def sink[T](partitioner: T ⇒ Int, startAfterNbrOfStreams: Int, bufferSize: Int = 256): Sink[T, Source[T, NotUsed]] =
+    Sink.fromGraph(new PartitionHub[T](partitioner, startAfterNbrOfStreams, bufferSize))
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private[akka] object Internal {
+    sealed trait ConsumerEvent
+    case object Wakeup extends ConsumerEvent
+    final case class HubCompleted(failure: Option[Throwable]) extends ConsumerEvent
+    case object Initialize extends ConsumerEvent
+
+    sealed trait HubEvent
+    case object RegistrationPending extends HubEvent
+    final case class UnRegister(id: Long) extends HubEvent
+    final case class NeedWakeup(consumer: Consumer) extends HubEvent
+    final case class Consumer(id: Long, callback: AsyncCallback[ConsumerEvent])
+    case object TryPull extends HubEvent
+
+    case object Completed
+
+    sealed trait HubState
+    final case class Open(callbackFuture: Future[AsyncCallback[HubEvent]], registrations: List[Consumer]) extends HubState
+    final case class Closed(failure: Option[Throwable]) extends HubState
+
+    private val FixedQueues = 128
+
+    // Need the queue to be pluggable to be able to use a more performant (less general)
+    // queue in Artery
+    trait PartitionQueue {
```

Comment:
Here are the benchmark results:
```
[info] Benchmark                                                           (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error   Units
[info] PartitionHubBenchmark.partition                                              256                  1  thrpt   12  4472452.035 ± 379272.863   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  1  thrpt   12     1214.771 ±    329.219  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  1  thrpt   12      287.892 ±     87.997    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  1  thrpt   12     1256.202 ±    434.457  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  1  thrpt   12      297.412 ±    108.877    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  1  thrpt   12        0.021 ±      0.019  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  1  thrpt   12        0.005 ±      0.005    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  1  thrpt   12       28.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  1  thrpt   12       23.000                   ms
[info] PartitionHubBenchmark.partition                                              256                  5  thrpt   12  4758452.104 ± 529411.579   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  5  thrpt   12      471.910 ±     48.670  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  5  thrpt   12      104.415 ±      0.849    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  5  thrpt   12      469.124 ±     91.323  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  5  thrpt   12      103.850 ±     18.231    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  5  thrpt   12        0.041 ±      0.031  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  5  thrpt   12        0.009 ±      0.007    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  5  thrpt   12       34.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  5  thrpt   12       20.000                   ms
[info] PartitionHubBenchmark.partition                                              256                 10  thrpt   12  4529333.257 ±  62294.979   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                 10  thrpt   12      460.360 ±      5.455  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                 10  thrpt   12      107.004 ±      0.127    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                 10  thrpt   12      469.674 ±     87.107  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                 10  thrpt   12      109.167 ±     20.081    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                 10  thrpt   12        0.039 ±      0.034  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                 10  thrpt   12        0.009 ±      0.008    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                 10  thrpt   12       33.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                 10  thrpt   12       21.000                   ms
```

and compared to when the `ConcurrentHashMap` is used (when id >= 128):
```
[info] Benchmark                                                           (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error   Units
[info] PartitionHubBenchmark.partition                                              256                  1  thrpt   12   991680.354 ± 100727.118   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  1  thrpt   12     1614.999 ±    219.292  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  1  thrpt   12     1710.444 ±     76.052    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  1  thrpt   12     1651.561 ±    270.543  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  1  thrpt   12     1749.595 ±    190.958    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  1  thrpt   12        0.052 ±      0.085  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  1  thrpt   12        0.055 ±      0.092    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  1  thrpt   12       53.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  1  thrpt   12       61.000                   ms
[info] PartitionHubBenchmark.partition                                              256                  5  thrpt   12  2638442.541 ±  79513.192   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                  5  thrpt   12     1006.254 ±     31.850  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                  5  thrpt   12      401.768 ±     20.979    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                  5  thrpt   12     1043.761 ±    320.936  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                  5  thrpt   12      417.359 ±    131.506    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                  5  thrpt   12        0.036 ±      0.055  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                  5  thrpt   12        0.014 ±      0.022    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                  5  thrpt   12       45.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                  5  thrpt   12       33.000                   ms
[info] PartitionHubBenchmark.partition                                              256                 10  thrpt   12  4088010.755 ±  69644.716   ops/s
[info] PartitionHubBenchmark.partition:·gc.alloc.rate                               256                 10  thrpt   12      438.160 ±      5.455  MB/sec
[info] PartitionHubBenchmark.partition:·gc.alloc.rate.norm                          256                 10  thrpt   12      112.839 ±      0.706    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space                      256                 10  thrpt   12      437.113 ±     76.592  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Eden_Space.norm                 256                 10  thrpt   12      112.608 ±     20.022    B/op
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space                  256                 10  thrpt   12        0.036 ±      0.037  MB/sec
[info] PartitionHubBenchmark.partition:·gc.churn.PS_Survivor_Space.norm             256                 10  thrpt   12        0.009 ±      0.009    B/op
[info] PartitionHubBenchmark.partition:·gc.count                                    256                 10  thrpt   12       30.000               counts
[info] PartitionHubBenchmark.partition:·gc.time                                     256                 10  thrpt   12       18.000                   ms
```

and compared to the `FixedSizePartitionHub` that is used in Artery, especially note the lower allocation rate:
```
[info] Benchmark                                                           (BufferSize)  (NumberOfStreams)   Mode  Cnt        Score        Error   Units
[info] PartitionHubBenchmark.arteryLanes                                            256                  1  thrpt   12  6889256.962 ± 531278.945   ops/s
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate                             256                  1  thrpt   12        8.263 ±      2.803  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate.norm                        256                  1  thrpt   12        1.283 ±      0.534    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space                    256                  1  thrpt   12       10.431 ±     31.206  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space.norm               256                  1  thrpt   12        1.568 ±      4.697    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space                256                  1  thrpt   12        0.281 ±      0.841  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space.norm           256                  1  thrpt   12        0.042 ±      0.127    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.count                                  256                  1  thrpt   12        2.000               counts
[info] PartitionHubBenchmark.arteryLanes:·gc.time                                   256                  1  thrpt   12       10.000                   ms
[info] PartitionHubBenchmark.arteryLanes                                            256                  5  thrpt   12  6059917.035 ± 498690.819   ops/s
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate                             256                  5  thrpt   12       34.218 ±      2.471  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate.norm                        256                  5  thrpt   12        5.948 ±      0.233    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space                    256                  5  thrpt   12       36.023 ±     40.771  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space.norm               256                  5  thrpt   12        6.390 ±      7.268    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space                256                  5  thrpt   12        0.956 ±      2.787  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space.norm           256                  5  thrpt   12        0.167 ±      0.487    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.count                                  256                  5  thrpt   12        7.000               counts
[info] PartitionHubBenchmark.arteryLanes:·gc.time                                   256                  5  thrpt   12       44.000                   ms
[info] PartitionHubBenchmark.arteryLanes                                            256                 10  thrpt   12  5634594.361 ±  83081.304   ops/s
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate                             256                 10  thrpt   12       53.854 ±      1.184  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.alloc.rate.norm                        256                 10  thrpt   12       10.062 ±      0.221    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space                    256                 10  thrpt   12       54.719 ±     22.260  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Eden_Space.norm               256                 10  thrpt   12       10.215 ±      4.159    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space                256                 10  thrpt   12        0.985 ±      2.856  MB/sec
[info] PartitionHubBenchmark.arteryLanes:·gc.churn.PS_Survivor_Space.norm           256                 10  thrpt   12        0.185 ±      0.537    B/op
[info] PartitionHubBenchmark.arteryLanes:·gc.count                                  256                 10  thrpt   12       11.000               counts
[info] PartitionHubBenchmark.arteryLanes:·gc.time                                   256                 10  thrpt   12       34.000                   ms
```